lib/Data/Upgrades/v3tov4.js

```diff
@@ -0,0 +1,94 @@
+import { cloneDeep } from 'lodash-es'
+import { formatCoordinate, oldBankIndexToCoordinate } from '../../Shared/ControlId.js'
+
+function CreateBankControlIdOld(page, bank) {
+	return `bank:${page}-${bank}`
+}
+
+function addControlIdsToPages(db) {
+	const controls = db.getKey('controls', {})
+	const pages = db.getKey('page', {})
+
+	const maxButtons = 32
+	const perRow = 8
+
+	for (const [number, page] of Object.entries(pages)) {
+		if (!page.controls) {
+			page.controls = {}
```

Comment:
I wonder if it would be better to continue the logic of the pages and split this in nested objects for rows and columns.
e.g. `"page":{"1":{"0": {"4": "bank:Xy3fZt4"}}}` for a control at row 0 and column 4 of page 1.

Reply:
A lot of the backend logic is using the 'coordinate' as a string, for simplicity of keeping it one parameter. which makes storing it in the db as a single level deep be simpler.

But I am wondering about whether instead of this string, whether doing it as an object would be messy. the page number could even be combined into that object when appropriate

Reply:
I think the most challenging part of this database is not storing data or deleting it (which is more mork with a nested structure), it is retrieving data. Today we have the need to get either the control on one specific coordinate, which is eqaully easy on all versions, or to get a complete page. I think in the future we will need an efficient way of retrieving a rectangular cutout. E.g for what is currently shown in UI or what is in the window.
Let's say we want all controls from 1/1 to 5/5. With the combined string we would need an Array.find() going over all the keys of the controls of the potentially huge page. Or we generate all existing coordinates in the range and test if there is a control at the position.
My mind's not made up for a specific solution yet, I'm just thinking loud.

Reply:
I don't think splitting it into another layer will help at all with that. Either way it will require the same level of searching.

But I am probably going to change this as part of other refactoring to stop storing a string coordinate.